Add V4L2_PIX_FMT_Y16_BE pixel format

firebladed opened the pull request:
Add V4L2_PIX_FMT_Y16_BE pixel format
#18495

swaps byte order then treats as V4L2_PIX_FMT_Y16
relates to https://github.com/opencv/opencv/pull/23944

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake
- [ ] Test using Melexis MLX90640

alalek commented on modules/videoio/src/cap_v4l.cpp:
```diff
@@ -1600,6 +1603,25 @@ void CvCaptureCAM_V4L::convertToRgb(const Buffer &currentBuffer)
+        };        
+        temp2.ForEach(<uint16_t>(Operator()); 
+        
+        temp2.convertTo(temp, CV_8U, 1.0 / 256);
```
Extraction through `cv::split()` from `CV_8UC2` should perform better.

alalek reviewed (commented):
Thank you for contribution!

Please make builds green.

asmorkalov commented:
@firebladed Do you have chance to finish the PR? Also it'll be great if you add list of know cameras that are enabled by the patch to the PR description. It becomes merge commit message and is very useful for git history.

firebladed commented:
@asmorkalov sorry, had issue an get it to work and been busy with other things recently, will try to look at again a soon as i can.

asmorkalov commented:
@firebladed Do you have chance to finish the patch? There are changes in GStreamer support that introduces conflicts.

firebladed commented:
I have a version that appears to work however i need to do some more tests and i don't know why the OpenCV CN Windows 10 x64 build failed

alalek commented:
> OpenCV CN

Ignore, there builders are optional for now.

firebladed commented:
Ive tested visually using the MLX90640 on ubuntu on a raspberypi and it appears to give the same output as
`ffmpeg -f rawvideo -pixel_format y16be -video_size "32x26" -i /dev/video0 -vf scale=288:234 -f fbdev -pix_fmt bgra /dev/fb0`

there appears to be some delay between running python script and getting an image which i cant currently account for

Id like to test using a stream that doesn't use the camera but so far i haven't found anything that can produce a raw y16be stream, ffmpeg doesn't appear to convert to y16be, that i can send to the videoloopback


For anyone wanting to use the MLX90640 ive also discovered that this method of obtaining video is of limited use, as the MLX90640 image correction requires data embedded in the video stream and implicit scaling to 8bit corrupts this data.    

have to set fourcc and disable rgb conversion before doing image correction and then convert to desired format
```
vs.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter.fourcc('Y','1','6',' '))
vs.set(cv2.CAP_PROP_CONVERT_RGB, 0)
```

asenyaev commented:
jenkins cn please retry a build

asmorkalov commented:
Thanks for the patch! OpenCV runs on Big-Endian platforms too (like PPC), so the patch with bytes swap does not work there. please add endianess check.

firebladed commented:
> Thanks for the patch! OpenCV runs on Big-Endian platforms too (like PPC), so the patch with bytes swap does not work there. please add endianess check.

sorry for slow reply

im not sure exactly what you mean, im also not familiar with PPC

unless im missing something

V4L2_PIX_FMT_Y16 and V4L2_PIX_FMT_Y16_BE

are the same except that the MSB and LSB are swapped for each pixel

so to convert between them i just swap the MSB and LSB, processor endianess is irrelevent

what should i be checking for endianess?

does 
```
       cv::split(temp,channels);
        cv::addWeighted(channels[0],256,channels[1],1,0,temp2,CV_16UC1);
```

not work on big endian platforms?

are you saying that i need to use this on big endian platforms?
```
       cv::split(temp,channels);
        cv::addWeighted(channels[1],256,channels[0],1,0,temp2,CV_16UC1);
```

mshabunin commented:
To check current platform you can copy this function to cap_v4l.cpp: https://github.com/opencv/opencv/blob/c4226f0457112ed6fc598d4b7456d130a1097085/modules/imgcodecs/src/utils.hpp#L136-L139

On BE platform we don't need to swap bytes of `V4L2_PIX_FMT_Y16_BE` image, while `V4L2_PIX_FMT_Y16` image should be transformed. So, perhaps your code should be extracted to a function and applied to either of the formats depending on the platform.

opencv-alalek commented:
According to documentation:
- https://www.kernel.org/doc/html/v4.10/media/uapi/v4l/pixfmt-y16-be.html
- https://www.kernel.org/doc/html/v4.10/media/uapi/v4l/pixfmt-y16.html

Implementation has been updated for using `cv::extractChannel()` call (no need to perform LE/BE checks for arithmetic operations)

mshabunin commented:
Maybe we should also add a new test case to the `videoio_v4l2.formats` with Y16_BE type?

mshabunin commented:
I've added this case to the test, it passes for me.

asmorkalov reviewed (approved):
:+1: 
Tested manually with Ubuntu 18.04 + vivid driver

opencv-alalek commented on modules/videoio/test/test_precomp.hpp:
```diff
@@ -60,7 +60,8 @@ inline std::string fourccToString(int fourcc)
     std::string res = fourccToString(fourcc);
-    std::replace_if(res.begin(), res.end(), [](uint8_t c){ return c < '0' || c > 'z'; }, '_');
+    // TODO: return hex values for invalid characters
+    std::transform(res.begin(), res.end(), res.begin(), [](uint8_t c) { return (c >= '0' && c <= 'z') ? c : (c == ' ' ? '_' : 'x'); });
```
There is build regression with MSVS2019/2017: https://pullrequest.opencv.org/buildbot/builders/4_x-win64-vc16/builds/100191

opencv-alalek replied:
Related CI issue: https://github.com/opencv/ci-gha-workflow/issues/108

mshabunin replied:
Created PR #24087
